File: OnBoard/Components/CanTp/Implementation/CanTp.c
#include "CanTp.h"
#include "CanIf.h"
#include "PduR.h"
#include "PduR_CanTp.h"
#include "ComStack_Types.h"
#include "Can_GeneralTypes.h"
#include "string.h"

#define CAN_STANDARD_SIZE 8

// TX
static const uint8_t *TX_appPtr = NULL;
static PduLengthType TX_totalLen = 0;
static PduLengthType TX_sentLen = 0;
static uint8_t TX_SN = 0;

static CanTpState txState = IDLE;

// RX
static PduLengthType RX_bufferSizePtr;
static uint8_t RX_SN = 0;
static PduLengthType totalLen;
static PduLengthType copiedLen;

Std_ReturnType CanTp_Transmit(PduIdType TxPduId, const PduInfoType *PduInfoPtr)
{
    if (!PduInfoPtr || !PduInfoPtr->SduDataPtr || PduInfoPtr->SduLength == 0u)
        return E_NOT_OK;
    if (txState != IDLE)
        return E_NOT_OK;

    if (PduInfoPtr->SduLength <= 7u) // Single Frame
    {
        PduInfoType info;
        static uint8_t sfBuf[8];
        info.SduDataPtr = sfBuf;
        info.MetaDataPtr = NULL;
        info.SduDataPtr[0] = (uint8_t)((PCI_TYPE_SF << 4) | (PduInfoPtr->SduLength & 0x0F));
        memcpy(&info.SduDataPtr[1], PduInfoPtr->SduDataPtr, PduInfoPtr->SduLength);
        info.SduLength = (PduLengthType)(1u + PduInfoPtr->SduLength);

        txState = SF;
        return CanIf_Transmit(TxPduId, &info);
    }

    if (PduInfoPtr->SduLength > 0x0FFFu)
        return E_NOT_OK;
    // Multiple Frame (First Frame)
    PduInfoType infoFirstFrame;
    static uint8_t ffBuf[8];

    infoFirstFrame.SduDataPtr = ffBuf;
    infoFirstFrame.MetaDataPtr = NULL;

    infoFirstFrame.SduDataPtr[0] = (uint8_t)((PCI_TYPE_FF << 4) | ((PduInfoPtr->SduLength >> 8) & 0x0F));
    infoFirstFrame.SduDataPtr[1] = (uint8_t)(PduInfoPtr->SduLength & 0xFF);

    memcpy(&infoFirstFrame.SduDataPtr[2], PduInfoPtr->SduDataPtr, 6u);
    infoFirstFrame.SduLength = 8u;

    TX_appPtr = PduInfoPtr->SduDataPtr;
    TX_totalLen = PduInfoPtr->SduLength;
    TX_sentLen = 6u;
    TX_SN = 1u;

    txState = FF;
    return CanIf_Transmit(TxPduId, &infoFirstFrame);
}

void CanTp_TxConfirmation(PduIdType TxPduId, Std_ReturnType result)
{
    if (result != E_OK)
    {
        txState = IDLE;
        PduR_CanTpTxConfirmation(TxPduId, result);
        return;
    }
    switch (txState)
    {
    case SF:
        txState = IDLE;
        PduR_CanTpTxConfirmation(TxPduId, result);
        break;

    case FF:
        txState = WAIT_FC;
        break;

    case CF: // Update: set flag & transmit on main function
    {
        PduLengthType left_before = (PduLengthType)(TX_totalLen - TX_sentLen);
        PduLengthType just = (left_before > 7u) ? 7u : left_before;
        TX_sentLen += just;
        TX_SN = (uint8_t)((TX_SN + 1u) & 0x0F);

        if (TX_sentLen >= TX_totalLen)
        {
            txState = IDLE;
            PduR_CanTpTxConfirmation(TxPduId, E_OK);
            break;
        }

        static uint8_t cfBuf[8];
        PduInfoType info;
        PduLengthType chunk = (PduLengthType)((TX_totalLen - TX_sentLen) > 7u ? 7u : (TX_totalLen - TX_sentLen));

        info.SduDataPtr = cfBuf;
        info.MetaDataPtr = NULL;
        info.SduDataPtr[0] = (uint8_t)((PCI_TYPE_CF << 4) | (TX_SN & 0x0F));
        memcpy(&info.SduDataPtr[1], &TX_appPtr[TX_sentLen], chunk);
        info.SduLength = (PduLengthType)(1u + chunk);

        if (CanIf_Transmit(TxPduId, &info) != E_OK)
        {
            txState = IDLE;
            PduR_CanTpTxConfirmation(TxPduId, E_NOT_OK);
        }
        break;
    }

    case WAIT_FC:
    case IDLE:
    default:
        break;
    }
}

void CanTp_MainFunction(void)
{
}

static void PrepareFC_AllowAll(PduIdType TxPduId)
{
    static uint8_t fc_buf[3];
    PduInfoType pdu;

    /* PCI: FT=FC(0x3) | FS=CTS(0x0) */
    fc_buf[0] = (uint8_t)((PCI_TYPE_FC << 4) | 0);
    fc_buf[1] = 0x00;
    fc_buf[2] = 0x00;

    pdu.SduDataPtr = fc_buf;
    pdu.MetaDataPtr = NULL;
    pdu.SduLength = (PduLengthType)sizeof(fc_buf); /* 3 byte */

    (void)CanIf_Transmit(TxPduId, &pdu);
}

void CanTp_RxIndication(PduIdType RxPduId, const PduInfoType *PduInfoPtr)
{
    PduLengthType TpSduLength = 0;
    PduInfoType info;
    BufReq_ReturnType bufferReq;
    switch ((PduInfoPtr->SduDataPtr[0] >> 4) & 0x0F)
    {
    case PCI_TYPE_SF:

        TpSduLength = PduInfoPtr->SduDataPtr[0] & 0x0F;
        if (TpSduLength == 0 || TpSduLength > 7 || TpSduLength > (PduInfoPtr->SduLength - 1))
        {
            return;
        }

        info.SduDataPtr = (uint8_t *)&PduInfoPtr->SduDataPtr[1];
        info.SduLength = TpSduLength;

        bufferReq = PduR_CanTpStartOfReception(RxPduId, &info, TpSduLength, &RX_bufferSizePtr);

        if (bufferReq != BUFREQ_OK || (RX_bufferSizePtr < info.SduLength))
        {
            return;
        }
        if (PduR_CanTpCopyRxData(RxPduId, &info, &RX_bufferSizePtr) == BUFREQ_OK)
        {

            PduR_CanTpRxIndication(RxPduId, E_OK);
        }

        break;

    case PCI_TYPE_FF:

        TpSduLength = ((PduInfoPtr->SduDataPtr[0] & 0x0F) << 8) | (PduInfoPtr->SduDataPtr[1]);
        if (TpSduLength == 0)
            return;

        PduLengthType firstChunk = (PduInfoPtr->SduLength > 2) ? (PduInfoPtr->SduLength - 2) : 0;
        if (firstChunk > 6)
            firstChunk = 6;
        info.SduDataPtr = (uint8_t *)&PduInfoPtr->SduDataPtr[2];
        info.MetaDataPtr = NULL;
        info.SduLength = firstChunk;

        bufferReq = PduR_CanTpStartOfReception(RxPduId, &info, TpSduLength, &RX_bufferSizePtr);

        if (bufferReq != BUFREQ_OK || RX_bufferSizePtr < info.SduLength)
            return;

        if (PduR_CanTpCopyRxData(RxPduId, &info, &RX_bufferSizePtr) == BUFREQ_OK)
        {
            PrepareFC_AllowAll(RxPduId);

            totalLen = TpSduLength;
            copiedLen = firstChunk;
            RX_SN = 0;
        }
        break;

    case PCI_TYPE_CF:
        if (totalLen == 0 || copiedLen >= totalLen) // Not received or already completed
        {
            PduR_CanTpRxIndication(RxPduId, E_NOT_OK);
            break;
        }

        uint8_t sn_expect = (uint8_t)((RX_SN + 1U) & 0x0Fu);
        uint8_t sn_rx = (uint8_t)(PduInfoPtr->SduDataPtr[0] & 0x0F);
        if (sn_rx != sn_expect) // Error sequence number
        {
            PduR_CanTpRxIndication(RxPduId, E_NOT_OK);
            break;
        }

        PduLengthType chunk = (PduInfoPtr->SduLength > 1u) ? (PduInfoPtr->SduLength - 1u) : 0u;
        PduLengthType left = (PduLengthType)(totalLen - copiedLen);
        if (chunk > left)
            chunk = left;

        PduInfoType info;
        info.SduDataPtr = (uint8_t *)&PduInfoPtr->SduDataPtr[1];
        info.MetaDataPtr = NULL;
        info.SduLength = chunk;

        if (PduR_CanTpCopyRxData(RxPduId, &info, &RX_bufferSizePtr) != BUFREQ_OK)
        {
            PduR_CanTpRxIndication(RxPduId, E_NOT_OK);
            break;
        }

        copiedLen = (PduLengthType)(copiedLen + chunk);
        RX_SN = sn_expect;

        if (copiedLen >= totalLen) // Reception complete
        {
            PduR_CanTpRxIndication(RxPduId, E_OK);
            totalLen = 0u;
            copiedLen = 0u;
            RX_SN = 0u;
        }
        break;
    case PCI_TYPE_FC:
        if (txState == WAIT_FC && PduInfoPtr->SduLength >= 3u)
        {
            uint8_t fs = (uint8_t)(PduInfoPtr->SduDataPtr[0] & 0x0F);

            if (fs == FC_FS_CTS)
            {
                static uint8_t cfBuf[8];
                PduInfoType info;

                PduLengthType left = (PduLengthType)(TX_totalLen - TX_sentLen);
                PduLengthType chunk = (left > 7u) ? 7u : left;

                info.SduDataPtr = cfBuf;
                info.MetaDataPtr = NULL;
                info.SduDataPtr[0] = (uint8_t)((PCI_TYPE_CF << 4) | (TX_SN & 0x0F));
                memcpy(&info.SduDataPtr[1], &TX_appPtr[TX_sentLen], chunk);
                info.SduLength = (PduLengthType)(1u + chunk);

                txState = CF;
                if (CanIf_Transmit(RxPduId, &info) != E_OK) // send first CF
                {
                    txState = IDLE;
                    PduR_CanTpTxConfirmation(RxPduId, E_NOT_OK);
                }
            }
            else if (fs == FC_FS_WT)
            {
            }
            else
            {
                txState = IDLE;
                PduR_CanTpTxConfirmation(RxPduId, E_NOT_OK);
            }
        }
        break;
    }
}

File: OnHost/bus_monitor/Components/CanTp/CanTp.py
from __future__ import annotations

from typing import Dict, List, Optional

from Components.models.can_frame import CanFrame
from Components.CanTp.CanTp_Types import (
    FrameType,
    CanTpState,
    CanTpFcFsType,
    CanTpChannelConfig,
)


class CanTp:
    def __init__(
        self,
        can_if,
        pdur,
        channels: List[CanTpChannelConfig],
    ) -> None:
        self._can_if = can_if
        self._pdur = pdur

        self._channels_by_rx_pdu_id: Dict[int, CanTpChannelConfig] = {
            ch.rx_pdu_id: ch for ch in channels
        }
        self._channels_by_tx_pdu_id: Dict[int, CanTpChannelConfig] = {
            ch.tx_pdu_id: ch for ch in channels
        }

        # -------------------------
        # TX context
        # -------------------------
        self._tx_app_ptr: bytes = b""
        self._tx_total_len: int = 0
        self._tx_sent_len: int = 0
        self._tx_sn: int = 0
        self._tx_state: CanTpState = CanTpState.IDLE
        self._tx_active_pdu_id: Optional[int] = None

        # -------------------------
        # RX context
        # -------------------------
        self._rx_total_len: int = 0
        self._rx_copied_len: int = 0
        self._rx_sn: int = 0
        self._rx_buffer: bytearray = bytearray()

    # --------------------------------------------------
    # Public API
    # --------------------------------------------------
    def CanTp_Transmit(self, tx_pdu_id: int, payload: bytes) -> bool:
        if payload is None or len(payload) == 0:
            return False

        if self._tx_state != CanTpState.IDLE:
            return False

        channel = self._channels_by_tx_pdu_id.get(tx_pdu_id)
        if channel is None:
            return False

        # Single Frame
        if len(payload) <= 7:
            sf = (
                bytes([(FrameType.PCI_TYPE_SF.value << 4) | (len(payload) & 0x0F)])
                + payload
            )

            self._tx_state = CanTpState.SF
            self._tx_active_pdu_id = tx_pdu_id

            return self._can_if.transmit(tx_pdu_id, sf)

        # First Frame
        if len(payload) > 0x0FFF:
            return False

        ff_byte0 = (FrameType.PCI_TYPE_FF.value << 4) | ((len(payload) >> 8) & 0x0F)
        ff_byte1 = len(payload) & 0xFF
        ff = bytes([ff_byte0, ff_byte1]) + payload[:6]

        self._tx_app_ptr = payload
        self._tx_total_len = len(payload)
        self._tx_sent_len = 6
        self._tx_sn = 1
        self._tx_active_pdu_id = tx_pdu_id
        self._tx_state = CanTpState.FF

        return self._can_if.transmit(tx_pdu_id, ff)

    def CanTp_MainFunction(self) -> None:
        # Bản đơn giản: chưa xử lý timer
        pass

    def CanTp_RxIndication(self, rx_pdu_id: int, frame: CanFrame) -> None:
        channel = self._channels_by_rx_pdu_id.get(rx_pdu_id)
        if channel is None:
            return

        if frame.data is None or len(frame.data) == 0:
            return

        pci_type = (frame.data[0] >> 4) & 0x0F

        # --------------------------------------------------
        # Single Frame
        # --------------------------------------------------
        if pci_type == FrameType.PCI_TYPE_SF.value:
            tp_len = frame.data[0] & 0x0F

            if tp_len == 0 or tp_len > 7:
                return

            if tp_len > (len(frame.data) - 1):
                return

            payload = bytes(frame.data[1 : 1 + tp_len])
            self._pdur.PduR_CanTpRxIndication(rx_pdu_id, payload, True)
            return

        # --------------------------------------------------
        # First Frame
        # --------------------------------------------------
        if pci_type == FrameType.PCI_TYPE_FF.value:
            if len(frame.data) < 2:
                return

            tp_len = ((frame.data[0] & 0x0F) << 8) | frame.data[1]
            if tp_len == 0:
                return

            first_chunk = frame.data[2:]
            if len(first_chunk) > 6:
                first_chunk = first_chunk[:6]

            self._rx_total_len = tp_len
            self._rx_buffer = bytearray(first_chunk)
            self._rx_copied_len = len(first_chunk)
            self._rx_sn = 0

            self._prepare_fc_allow_all(channel.fc_tx_pdu_id)
            return

        # --------------------------------------------------
        # Consecutive Frame
        # --------------------------------------------------
        if pci_type == FrameType.PCI_TYPE_CF.value:
            if self._rx_total_len == 0 or self._rx_copied_len >= self._rx_total_len:
                self._pdur.PduR_CanTpRxIndication(rx_pdu_id, b"", False)
                return

            sn_expect = (self._rx_sn + 1) & 0x0F
            sn_rx = frame.data[0] & 0x0F

            if sn_rx != sn_expect:
                self._reset_rx_context()
                self._pdur.PduR_CanTpRxIndication(rx_pdu_id, b"", False)
                return

            chunk = frame.data[1:]
            left = self._rx_total_len - self._rx_copied_len
            if len(chunk) > left:
                chunk = chunk[:left]

            self._rx_buffer.extend(chunk)
            self._rx_copied_len += len(chunk)
            self._rx_sn = sn_expect

            if self._rx_copied_len >= self._rx_total_len:
                payload = bytes(self._rx_buffer[: self._rx_total_len])
                self._pdur.PduR_CanTpRxIndication(rx_pdu_id, payload, True)
                self._reset_rx_context()

            return

        # --------------------------------------------------
        # Flow Control
        # --------------------------------------------------
        if pci_type == FrameType.PCI_TYPE_FC.value:
            if self._tx_state == CanTpState.WAIT_FC and len(frame.data) >= 3:
                fs = frame.data[0] & 0x0F

                if fs == CanTpFcFsType.FC_FS_CTS.value:
                    self._send_first_cf_after_fc(rx_pdu_id)
                    return

                if fs == CanTpFcFsType.FC_FS_WT.value:
                    return

                self._tx_abort(False)

    def CanTp_TxConfirmation(self, tx_pdu_id: int, result: bool = True) -> None:
        if not result:
            self._tx_abort(False)
            return

        if self._tx_state == CanTpState.SF:
            self._tx_state = CanTpState.IDLE
            self._pdur.PduR_CanTpTxConfirmation(tx_pdu_id, True)
            return

        if self._tx_state == CanTpState.FF:
            self._tx_state = CanTpState.WAIT_FC
            return

        if self._tx_state == CanTpState.CF:
            left_before = self._tx_total_len - self._tx_sent_len
            just = 7 if left_before > 7 else left_before
            self._tx_sent_len += just
            self._tx_sn = (self._tx_sn + 1) & 0x0F

            if self._tx_sent_len >= self._tx_total_len:
                self._tx_state = CanTpState.IDLE
                self._pdur.PduR_CanTpTxConfirmation(tx_pdu_id, True)
                self._reset_tx_context()
                return

            chunk = self._tx_app_ptr[self._tx_sent_len : self._tx_sent_len + 7]
            cf = (
                bytes([(FrameType.PCI_TYPE_CF.value << 4) | (self._tx_sn & 0x0F)])
                + chunk
            )

            if not self._can_if.transmit(tx_pdu_id, cf):
                self._tx_abort(False)

    # --------------------------------------------------
    # Internal helpers
    # --------------------------------------------------
    def _prepare_fc_allow_all(self, tx_pdu_id: int) -> None:
        fc = bytes(
            [
                (FrameType.PCI_TYPE_FC.value << 4) | CanTpFcFsType.FC_FS_CTS.value,
                0x00,
                0x00,
            ]
        )
        self._can_if.transmit(tx_pdu_id, fc)

    def _send_first_cf_after_fc(self, tx_pdu_id: int) -> None:
        left = self._tx_total_len - self._tx_sent_len
        chunk_len = 7 if left > 7 else left

        cf = (
            bytes([(FrameType.PCI_TYPE_CF.value << 4) | (self._tx_sn & 0x0F)])
            + self._tx_app_ptr[self._tx_sent_len : self._tx_sent_len + chunk_len]
        )

        self._tx_state = CanTpState.CF

        if not self._can_if.transmit(tx_pdu_id, cf):
            self._tx_abort(False)

    def _reset_rx_context(self) -> None:
        self._rx_total_len = 0
        self._rx_copied_len = 0
        self._rx_sn = 0
        self._rx_buffer = bytearray()

    def _reset_tx_context(self) -> None:
        self._tx_app_ptr = b""
        self._tx_total_len = 0
        self._tx_sent_len = 0
        self._tx_sn = 0
        self._tx_active_pdu_id = None

    def _tx_abort(self, result: bool) -> None:
        active = self._tx_active_pdu_id
        self._tx_state = CanTpState.IDLE
        self._reset_tx_context()

        if active is not None:
            self._pdur.PduR_CanTpTxConfirmation(active, result)
